src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2058,17 +2057,27 @@ class ShenandoahUpdateHeapRefsTask : public WorkerTask {
     if (CONCURRENT) {
       ShenandoahConcurrentWorkerSession worker_session(worker_id);
       ShenandoahSuspendibleThreadSetJoiner stsj;
-      do_work<ShenandoahConcUpdateRefsClosure>();
+      do_work<ShenandoahConcUpdateRefsClosure>(worker_id);
     } else {
       ShenandoahParallelWorkerSession worker_session(worker_id);
-      do_work<ShenandoahSTWUpdateRefsClosure>();
+      do_work<ShenandoahSTWUpdateRefsClosure>(worker_id);
     }
   }
 
 private:
   template<class T>
-  void do_work() {
+  void do_work(uint worker_id) {
     T cl;
+    if (CONCURRENT && (worker_id == 0)) {
```

Comment:
Could move this work to happen on the control thread? Not sure if there is an advantage to having a single worker do this. Does it matter when other workers observe the effect of `move_collector_sets_to_mutator`?

Reply:
Was thinking it's best to let the other N-1 worker threads get right to work on updating refs, while this first worker thread starts moving the available collector regions to mutator free set

Reply:
We want to put these Collector regions into the service of the Mutator threads as quickly as possible in order to avoid Mutator allocation failures.  We also want to complete update-refs as quickly as possible, because that will put a lot more memory into the hands of the Mutators.

The implementation of move_collector_sets_to_mutator() takes the heap lock when it reassigns a region's purpose from Collector to Mutator.

Reply:
Okay.

Reply:
Cobbling together update-refs and freeset management adds to already high complexity. We need to do this as separate task somewhere in `op_updaterefs()`. Also, what does this move for _non-concurrent_ update refs, like degen GC?